compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt

```diff
@@ -277,43 +279,63 @@ private class BridgeLowering(val context: JvmBackendContext) : ClassLoweringPass
+        // code directly as a prelude in the existing method.
+        val variableMap = mutableMapOf<IrValueParameter, IrValueParameter>()
+        if (specialOverrideSignature == ourSignature) {
+            val argumentsToCheck = valueParameters.take(specialOverrideInfo.argumentsToCheck)
+            val shouldGenerateParameterChecks = argumentsToCheck.any { !it.type.isNullable() }
+            if (shouldGenerateParameterChecks) {
+                // Rewrite the body to check if arguments have wrong type. If so, return the default value, otherwise,
+                // use the existing function body.
+                context.createIrBuilder(symbol).run {
+                    body = irBlockBody {
+                        // Change the parameter types to be Any? so that null checks are not generated. The checks
+                        // we insert here make them superfluous.
+                        argumentsToCheck.forEach {
+                            val parameterType = it.type
+                            if (!parameterType.isNullable()) {
+                                val newParameter = it.copyTo(this@rewriteSpecialMethodBody, type = context.irBuiltIns.anyNType)
+                                variableMap.put(valueParameters[it.index], newParameter)
+                                valueParameters[it.index] = newParameter
+                                addParameterTypeCheck(
+                                    newParameter,
+                                    parameterType,
+                                    specialOverrideInfo.defaultValueGenerator,
+                                    this@rewriteSpecialMethodBody
+                                )
+                            }
+                        }
+                        // After the checks, insert the orignal method body.
+                        if (body is IrExpressionBody) {
+                            +(body as IrExpressionBody).expression
+                        } else {
+                            (body as IrBlockBody).statements.forEach { +it }
                         }
-                    }
-
-                    // Map usages of rewritten parameters in old body to the new parameters and paste in the body after the explicit
-                    // null check.
-                    body?.transform(VariableRemapper(variableMap), null)
-                    if (body is IrExpressionBody) {
-                        +(body as IrExpressionBody).expression
-                    } else {
-                        (body as IrBlockBody).statements.forEach { +it }
                     }
                 }
             }
+        } else {
+            // If the signature of this method will be changed in the output to take a boxed argument instead of a primitive,
+            // rewrite the argument so that code will be generated for a boxed argument and not a primitive.
```

Comment:
I would prefer a more explicit way to indicate that an argument should not be realized as a primitive type, but I guess `makeNullable()` will do for now.